HOSTEDCP-445: Documented the possible issues found during a HostedCluster migration (#2276)
openshift/hypershift: docs/content/how-to/aws/troubleshooting.md

---
title: Troubleshooting
---

# Troubleshooting

## Disaster Recovery - HostedCluster Migration

These are issues related with disaster recovery that we've identified and you could face during a HostedCluster migration.

> ### The new workloads does not got scheduled in the new migrated cluster

Reviewer: ```suggestion
### The new workloads do not got scheduled in the new migrated cluster
```

> Everything looks normal, in the destination Management or Hosted Cluster and in the old Management and Hosted Cluster, but your new workloads does not schedulles in your migrated Hosted Cluster (your old ones should work properly).

Reviewer: ```suggestion
Everything looks normal, in the destination Management or Hosted Cluster and in the old Management and Hosted Cluster, but your new workloads do not schedule in your migrated Hosted Cluster (your old ones should work properly).
```

Eventually your pods begin to fall down and the cluster status becomes degraded.

1. First thing you need to check is the cluster operators and validate all of them work properly:
```
oc get co
```

2. If there are some of them degraded and with errors, please check the logs and validate things point to an OVN issue.
```
oc get co <Operator's Name> -o yaml
```

3. To solve the issue, we need to ensure the old Hosted Cluster is in pause ([we also need this](https://github.com/openshift/hypershift/pull/2265)) or deleted.

4. Now we need to delete the OVN pods
```
oc --kubeconfig=${HC_KUBECONFIG} delete pod -n openshift-ovn-kubernetes --all
```

Eventually the Hosted Cluster will start self healing and the ClusterOperator will come back.

**Cause:** The cause of this issue is after the Hosted Cluster Migration the KAS (Kube API Server) uses the same DNS name, but it points to different load balancer in AWS platform. Sometimes OVN does not behave correctly facing this situation.

### The migration gets blocked in ETCD recovery

The context around it's basically "I've edited the Hosted Cluster adding the `ETCDSnapshotURL` but the modification disappears and does not continue".

> The first simptom is the status of the Hypershift Operator pod, in this case the pod is usually in `CrashLoopBackOff` status.

Reviewer 2: nit: s/simptom/symptom/

To solve this issue we need to:

1. Kill the Hypershift Operator pod
```
oc delete pod -n hypershift -lapp=operator
```

> 2. Continue editing the pod, in order to add the `ETCDSnapshotURL` to the Hosted Cluster spec.

Reviewer 2: Why would you edit the pod, shouldn't you edit the hosted cluster?

Author: yep, you're right, In my mind was also edit HostedCluster xD

3. Now the ETCD pod will raise up using the snapshot from S3 bucket.

> **Cause:** This issue happens when the Hypershift operator is down and the Hosted Cluster controller cannot handle the modifications in the objects which belong to him.

Reviewer 2: s/which belong to him/which belong to it/

### The nodes cannot join the new Hosted Cluster and stay in the older one

We have 2 paths to follow and depends if [this code](https://github.com/openshift/hypershift/pull/2265) is in your Hypershift Operator.

#### The PR is merged and my Hypershift Operator has that code running

If that's the case, you need to make sure your Hosted Cluster is paused:
```
oc get hostedcluster -n <HC Namespace> <HC Name> -ojsonpath={.Spec.pausedUntil}
```

If this command does not give you any output, make sure you've followed properly the "[Disaster Recovery](https://hypershift-docs.netlify.app/how-to/aws/disaster-recovery/)" procedure, more concretelly pausing the Hosted Cluster and NodePool.

Even if it's paused and is still in that situation, please **continue to the next section** because it's highly probable that you don't have the code which manages this situation properly.

#### The PR is not merged or my Hypershift Operator does not have that code running

If that's not the case, the only way to solve it is executing the teardown of the old Hosted Cluster prior the full restoration in the new Management cluster. Make sure you already have all the Manifests and the ETCD backed up.

> Once you followed the Teardown procedure of the old Hosted Cluster, you will see how the migrated Hosted Cluster begin to self-recover.

Reviewer 2: s/begin/begins/

> Once you followed the Teardown procedure of the old HostedCluster, you will see how the migrated HostedCluster begins to self-recover

Reviewer: ```suggestion
Once you followed the Teardown procedure of the old HostedCluster, you will see how the migrated HostedCluster begins to self-recover.
```

> **Cause:** This issue occurs when the old HostedCluster have a conflict with the AWSPrivateLink object. The old one is still running and the new one cannot handle it because the `hypershift.local` AWS internal DNS entry still points to the old LoadBalancer.

Reviewer: ```suggestion
**Cause:** This issue occurs when the old HostedCluster has a conflict with the AWSPrivateLink object. The old one is still running and the new one cannot handle it because the `hypershift.local` AWS internal DNS entry still points to the old LoadBalancer.
```

> ### Dependant resources blocks the old HostedCluster teardown

Reviewer: ```suggestion
### Dependent resources block the old HostedCluster teardown
```

> To solve this issue you need to check all the objects in the HostedControlPlane Namespace and make sure all of then are being terminated. To do that we recommend to use an external tool called [ketall](https://github.com/corneliusweig/ketall) which gives you a complete overview of all resources in a kubernetes cluster.

Reviewer: ```suggestion
To solve this issue, you need to check all the objects in the HostedControlPlane Namespace and make sure all of them are being terminated. To do that, we recommend to use an external tool called [ketall](https://github.com/corneliusweig/ketall) which gives you a complete overview of all resources in a kubernetes cluster.
```

> You need to know what object is preventing that the HostedCluster gets deleted and ensure that the finalizer finishes successfully.

Reviewer: ```suggestion
You need to know what object is preventing the HostedCluster from being deleted and ensure that the finalizer finishes successfully.
```

If you don't care about the stability of the old Management cluster, this script could help you to delete all the components in the **HostedControlPlane Namespace** (you need the ketall tool):

```
#!/bin/bash

####
# Execution sample:
# ./delete_ns.sh $NAMESPACE
####

NAMESPACE=$1

if [[ -z $1 ]];then
        echo "Specify the Namespace!"
        exit 1
fi

for object in $(ketall -n $NAMESPACE -o name | grep -v packa)
do
    oc -n $NAMESPACE patch $object -p '{"metadata":{"finalizers":null}}' --type merge
done
```

Eventually, the namespace will be successfully terminated and also the HostedCluster.

> **Cause:** This is pretty common issue in the Kubernetes/Openshift world. You are trying to delete a resource that it has other dependedant objects. The finalizer is still trying to delete them but it cannot progress.

Reviewer: ```suggestion
**Cause:** This is pretty common issue in the Kubernetes/Openshift world. You are trying to delete a resource that has other dependent objects. The finalizer is still trying to delete them but it cannot progress.
```
